go/arrow/ipc/compression.go

```diff
@@ -0,0 +1,106 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipc
+
+import (
+	"io"
+
+	"github.com/apache/arrow/go/arrow/internal/flatbuf"
+	"github.com/klauspost/compress/zstd"
+	"github.com/pierrec/lz4/v4"
+)
+
+type compressor interface {
+	MaxCompressedLen(n int) int
+	Reset(io.Writer)
+	io.WriteCloser
+	Type() flatbuf.CompressionType
+}
+
+type lz4Compressor struct {
+	*lz4.Writer
+}
+
+func (lz4Compressor) MaxCompressedLen(n int) int {
+	return lz4.CompressBlockBound(n)
+}
+
+func (lz4Compressor) Type() flatbuf.CompressionType {
+	return flatbuf.CompressionTypeLZ4_FRAME
+}
+
+type zstdCompressor struct {
+	*zstd.Encoder
+}
+
+// from zstd.h
+func (zstdCompressor) MaxCompressedLen(len int) int {
+	extra := ((128 << 10) - len) >> 11
```

Comment:
please be careful here with signed values here.   it looks like ZSTD_compressBound takes size_t which I believe is unsigned.

Reply:
updated to use `uint` here to ensure that we don't make any mistakes with signed vs unsigned values.